### PR TITLE
Enable secrets in federation kubectl tests

### LINKED ISSUE #40568: enable kubectl tests with federation apiserver for secrets
https://github.com/kubernetes/kubernetes/pull/40567 disabled them since `kubectl delete secrets` was flaky.

We need to find and fix the issue and then reenable these tests.

cc @kubernetes/sig-federation-bugs 

Related issues: #39984 and #39985

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/kubernetes/kubernetes/issues/40568
Superseedes https://github.com/kubernetes/kubernetes/pull/40714

Updating kubectl tests to wait for deletion if WAIT_FOR_DELETION is set to true. WAIT_FOR_DELETION will be set to true only when the tests are being run for federation apiserver.
This change will not impact kube apiserver tests and still enable federation and kubernetes to share the same test code.
This is a workaround until https://github.com/kubernetes/kubernetes/issues/42594 is fixed.

cc @kubernetes/sig-federation-pr-reviews
cc @liggitt as he reviewed https://github.com/kubernetes/kubernetes/pull/40714

### CODE REVIEW

File: hack/make-rules/test-cmd-util.sh [hunk 1]
```diff
@@ -1813,7 +1813,9 @@ run_secrets_test() {
 
   ### Create a docker-registry secret in a specific namespace
   # Pre-condition: no SECRET exists
-  kube::test::get_object_assert 'secrets --namespace=test-secrets' "{{range.items}}{{$id_field}}:{{end}}" ''
```
Comment (liggitt) [inline hack/make-rules/test-cmd-util.sh]: this looks like it is removing the precondition checks entirely for non-federation runs

Comment (nikhiljindal) [inline hack/make-rules/test-cmd-util.sh]: oops, I feel silly!
fixed. thanks for the catch.

Comment (liggitt) [inline hack/make-rules/test-cmd-util.sh]: I would leave the existing code alone and add this before it:
```
  if [[ "${WAIT_FOR_DELETION}" == "true" ]]; then
    kube::test::wait_object_assert 'secrets --namespace=test-secrets' "{{range.items}}{{$id_field}}:{{end}}" ''
  fi
  # Pre-condition: no SECRET exists
  kube::test::get_object_assert 'secrets --namespace=test-secrets' "{{range.items}}{{$id_field}}:{{end}}" ''
```

that makes it super clear that the precondition always holds, and makes removing the wait very clear and separate

Comment (nikhiljindal) [inline hack/make-rules/test-cmd-util.sh]: Done

**Comment by nikhiljindal** [comment]
Thanks for the review @liggitt Updated code as per comments.
PTAL

**Comment by nikhiljindal** [comment]
Adding 1.6 milestone which should be fine since this is a test only change.

**Comment by liggitt** [comment]
> Adding 1.6 milestone which should be fine since this is a test only change.

Are the federation tests queue blocking? Does this have the potential to introduce flakes in secret deletion tests?

**Comment by liggitt** [comment]
change LGTM, though I question it being in the 1.6 milestone
please add a task in https://github.com/kubernetes/kubernetes/issues/42594 to revert this

**Comment by nikhiljindal** [comment]
Yes these tests are merge queue blocking.
I have run them locally and ensured that they are not flaky.

These are not really required for 1.6. Just good to have. More tests add confidence in the release :)

**Comment by nikhiljindal** [comment]
Added a task to #42594

**Comment by liggitt** [comment]
at this point, v1.6 is supposed to mean we wouldn't ship without it. LGTM but for 1.7

**Comment by nikhiljindal** [comment]
sure thanks @liggitt 
Assigning to @ixdy for hack/OWNERS approval

**Comment by nikhiljindal** [comment]
@k8s-bot bazel test this

**Pull request merged**